src/debug.cc

```diff
@@ -91,6 +372,214 @@ ResyncDebugLog(FILE *newFile)
+    (void)module.cacheLogChannel.releaseEarlyMessages();
+    (void)module.stderrChannel.releaseEarlyMessages();
+    (void)module.syslogChannel.releaseEarlyMessages();
+}
+
+void
+Debug::SwanSong()
+{
+    Module().swanSong();
+}
+
+void
+DebugChannel::logSaved(const DebugMessages &messages)
+{
+    const auto writtenEarlier = written;
+    for (const auto &message: messages) {
+        if (Debug::Enabled(message.header.section, message.header.level) &&
+                lastWrittenRecordNumber < message.header.recordNumber)
+            log(message.header, message.body);
+    }
+    const auto writtenNow = written - writtenEarlier;
+    debugs(0, 5, "wrote " << writtenNow << " out of " <<
+           messages.size() << " early messages to " << name);
+}
+
+bool
+DebugChannel::saveMessage(const DebugMessageHeader &header, const std::string &body)
+{
+    if (!earlyMessages)
+        return false; // we have stopped saving early messages
+
+    if (header.level > EarlyMessagesLevel)
+        return false; // this message is not important enough to save
+
+    // There should not be a lot of messages because EarlyMessagesLevel is
+    // small, but we limit their accumulation just in case.
+    const DebugMessages::size_type limit = 1000;
+    DebugMessages::size_type purged = 0;
+    if (earlyMessages->size() >= limit) {
+        DebugMessages doomedMessages;
+        earlyMessages->swap(doomedMessages);
+        handleOverflow(doomedMessages);
+        purged = doomedMessages.size();
+    }
+
+    earlyMessages->emplace_back(header, body);
+
+    if (purged) {
+        // This debugs() should come _after_ we save the early message above,
+        // preserving the original event and LogMessage() order, like log().
+        debugs(0, DBG_CRITICAL, "ERROR: Too many early important messages. " <<
+               "Purged " << purged << " from " << name);
+    }
+
+    return true;
+}
+
+void
+DebugChannel::writeToStream(FILE &destination, const DebugMessageHeader &header, const std::string &body)
+{
+    fprintf(&destination, "role=%s # %s%s| %s\n",
+            header.role_XXX,
+            debugLogTime(header.timestamp),
+            debugLogKid(),
+            body.c_str());
+    noteWritten(header);
+}
+
+void
+DebugChannel::noteWritten(const DebugMessageHeader &header)
+{
+    ++written;
+    lastWrittenRecordNumber = header.recordNumber;
+}
+
+/* CacheLogChannel */
+
+void
+CacheLogChannel::log(const DebugMessageHeader &header, const std::string &body)
+{
+    assert(header.recordNumber > lastWrittenRecordNumber);
+
+    if (earlyMessages)
+        return (void)saveMessage(header, body);
+
+    if (!TheLog.file())
+        return;
+
+    writeToStream(*TheLog.file(), header, body);
+    fflush(TheLog.file());
+}
+
+void
+CacheLogChannel::handleOverflow(DebugMessages &doomedMessages)
+{
+    for (auto &message: doomedMessages)
+        message.header.overflowed = true;
+    Module().stderrChannel.logSaved(doomedMessages);
```

Comment:
Is there a possibility that stderrChannel logs a message twice, i.e., if it belongs to stderrChannel and has been ready stored in stderrChannel.earlyMessages? Or the 'lastWrittenRecordNumber' check in DebugChannel::logSaved() will protect from this?